File: leetcode_with_adam/docs/leetcode/605-can-place-flowers/605.js
/**
 * @param {number[]} flowerbed
 * @param {number} n
 * @return {boolean}
 */
var canPlaceFlowers = function(flowerbed, n) {
    // create a larger array and prepend and append 0, size is 2 + flowerbed.length
    // iterate throught the index 1 to 1 + flowerbed.length,  
    // - if 1, we skip the rest of the iteration
    // - else, we check for left,right condition if yes, we increment the counter 
    //
    const bigger_flowerbed = [0];
    bigger_flowerbed.push(...flowerbed);
    bigger_flowerbed.push(0);

    var counter = 0;
    for (let i = 0; i < 1 + flowerbed.length; i++) {
        if (bigger_flowerbed[i] === 1) {
            continue;
        }

        if (bigger_flowerbed[i-1] == 0 && bigger_flowerbed[i+1] == 0) {
            counter += 1;
            bigger_flowerbed[i] = 1;
        }
     }
    
    console.log(bigger_flowerbed);
    console.log(counter);
    return true ? counter >= n : false
};

const flowerbed = [1,0,0,0,1];
const n = 1;

canPlaceFlowers(flowerbed,n);

File: leetcode_with_adam/docs/leetcode/605-can-place-flowers/605.java
class Solution {
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        // initialize an array of zero with 2 unit larger than flowerbed
        // grab it
        // 
        int[] biggerFlowerbed = new int[flowerbed.length + 2];
        
        for (int i = 1; i < 1 + flowerbed.length; i++) {
            biggerFlowerbed[i] = flowerbed[i-1];
        }
        int counter = 0;
        // if left is 1, then we don't plant
        for (int i = 1; i < biggerFlowerbed.length - 1; i++) {
            // planted
            if (biggerFlowerbed[i] == 1) {
                continue;
            }
            // check if neighbors are empty (eligible for plant)
            if (biggerFlowerbed[i-1] == 0 && biggerFlowerbed[i+1] == 0) {
                counter++;
                biggerFlowerbed[i] = 1;
            } 
        }
        if (counter >= n) {
            return true;
        }
        return false;
    }

    public static void main(String[] args) {
        // 1. migrate the flowerbed --> bigger flowerbed with dummy header and dummy tail just like
        // how we uniformally treat insert/delete for linked list
        //
        Solution solution = new Solution(); 
        int[] flowerbed = {1,0,0,0,1};
        int n = 1;

        boolean res = solution.canPlaceFlowers(flowerbed,n);
    }
}
